include/albatross/src/models/pic_gp.hpp

```diff
@@ -0,0 +1,845 @@
+    std::vector<std::size_t> feature_to_block(features.size());
+    std::vector<decltype(sparse_gp_fit.measurement_groups.begin())> groups(
+        features.size());
+    Eigen::VectorXi col_alloc{Eigen::VectorXi::Zero(features.size())};
+    bool all_same_group = true;
+    for (Eigen::Index j = 0; j < features.size(); ++j) {
+      groups[j] = sparse_gp_fit.measurement_groups.find(
+          independent_group_function_(without_measurement(features[j])));
+      if (groups[j] != sparse_gp_fit.measurement_groups.end()) {
+        col_alloc(j) = groups[j]->second.block_size;
+        all_same_group = all_same_group && groups[j] == groups[0];
+      }
+      feature_to_block[j] =
+          std::distance(sparse_gp_fit.measurement_groups.begin(), groups[j]);
+    }
+
+    // Second pass: compute mean vector and fill sparse Vp matrix
+    Eigen::VectorXd mean_correction = Eigen::VectorXd::Zero(features.size());
+    Eigen::SparseMatrix<double> Vp(sparse_gp_fit.train_features.size(),
+                                   features.size());
+    Vp.reserve(col_alloc);
+    for (Eigen::Index j = 0; j < features.size(); ++j) {
+      if (groups[j] == sparse_gp_fit.measurement_groups.end()) {
+        continue;
+      }
+
+      const auto &B = groups[j]->second;
+      Eigen::VectorXd Vbp =
+          this->covariance_function_(B.dataset.features,
+                                     std::vector<FeatureType>{features[j]}) -
+          K_up.col(j).transpose() * sparse_gp_fit.covariance_Y[B.block_index];
+      for (Eigen::Index i = 0; i < Vbp.size(); ++i) {
+        Vp.insert(B.initial_row + i, j) = Vbp(i);
+      }
+      mean_correction[j] = Vbp.dot(sparse_gp_fit.mean_w[B.block_index]);
+    }
+    Vp.makeCompressed();
+
+    Eigen::MatrixXd xi_lambda = sparse_gp_fit.A_ldlt.sqrt_solve(Vp);
+    Eigen::MatrixXd xi_u = sparse_gp_fit.Z * Vp;
+    Eigen::VectorXd VSV_diag{
+        (xi_lambda.transpose() * xi_lambda - xi_u.transpose() * xi_u)
+            .diagonal()};
+
+    const Eigen::VectorXd U_diag =
+        (K_up.transpose() * sparse_gp_fit.W * Vp).diagonal();
+
+    Eigen::VectorXd marginal_variance(cast::to_index(features.size()));
+    for (Eigen::Index i = 0; i < marginal_variance.size(); ++i) {
+      marginal_variance[i] = this->covariance_function_(
+          features[cast::to_size(i)], features[cast::to_size(i)]);
+    }
+
+    const Eigen::MatrixXd Q_sqrt =
+        sparse_gp_fit.train_covariance.sqrt_solve(K_up);
+    const Eigen::VectorXd Q_diag =
+        Q_sqrt.cwiseProduct(Q_sqrt).array().colwise().sum();
+    marginal_variance -= Q_diag;
+
+    const Eigen::MatrixXd S_sqrt =
+        sqrt_solve(sparse_gp_fit.sigma_R, sparse_gp_fit.P, K_up);
+    const Eigen::VectorXd S_diag =
+        S_sqrt.cwiseProduct(S_sqrt).array().colwise().sum();
+    marginal_variance += S_diag;
+
+    mean += mean_correction;
+
+    return MarginalDistribution(mean,
+                                marginal_variance - (2 * U_diag + VSV_diag));
```

Comment:
:clap: can't say I follow all of it, but I know this was a tough one. Well done.